### PR TITLE
condense init parameters

### PR DESCRIPTION
**Pull request by tomMoral**

### CODE REVIEW

**Comment by vloison** [comment]
Done: 
- Remove unused parameters in `FaDIn` init
- Condense remaining parameters in `FaDIn` init
- Update references to the `FaDIn` solver with new parameters
- Create custom solvers `FaDIn_no_precomputations` and `FaDIn_loglikelihood` to reproduce FaDIn experiments -> I couldn't run the experiments as I don't have tick, @GuillaumeStaermanML or @tomMoral could you please check that they still run?
- Add moment matching to FaDIn

File: fadin/loss_and_gradient.py [hunk 1]
```diff
@@ -1,6 +1,97 @@
...
+    with l2 loss and no precomputations."""
+    intens = solver.kernel_model.intensity_eval(
+        solver.params_intens[0],
+        solver.params_intens[1],
+        solver.params_intens[2:],
+        events_grid,
+        discretization
+    )
+    loss = discrete_l2_loss_conv(intens, events_grid, solver.delta)
+    loss.backward()
```
Comment (tomMoral) [inline fadin/loss_and_gradient.py:76]: Move this to `run_comp_autodiff.py`

**Review by tomMoral — COMMENTED** [review]
A first round of reviews

File: fadin/tests/test_loss_and_gradient.py [hunk 2]
```diff
@@ -207,7 +207,7 @@ def test_l2loss():
                                                       kernel_RC, n_events,
                                                       delta, end_time)
 
-    assert torch.isclose(loss_conv_RC, loss_precomp_RC)
+    assert torch.isclose(1 + loss_conv_RC, 1 + loss_precomp_RC)
```
[the remark this answers is not in the document]
Comment (vloison) [reply fadin/tests/test_loss_and_gradient.py:210]: When `torch.isclose` arguments are close to zero, an error can occur. Changint this line fixed failing tests from before. I added a comment to explicit this :)

File: fadin/solver.py [hunk 3]
```diff
@@ -413,7 +374,22 @@ def fit(self, events, end_time):
         return self
 
 
-def plot(solver, plotfig=False, bl_noise=False, title=None, ch_names=None, savefig=None):
+class FaDIn_no_precomputations(FaDIn):
```
[the remark this answers is not in the document]
Comment (vloison) [reply fadin/solver.py]: Thanks for the info about camelcase and snake case :)

File: fadin/solver.py [hunk 4]
```diff
@@ -302,78 +323,20 @@ def fit(self, events, end_time):
...
-                                                          self.params_intens[2:],
-                                                          events_grid,
-                                                          discretization)
-                if self.criterion == 'll':
-                    loss = discrete_ll_loss_conv(intens, events_grid, self.delta)
-                else:
-                    loss = discrete_l2_loss_conv(intens, events_grid, self.delta)
-                loss.backward()
-
+            self.optim_iteration(
```
[the remark this answers is not in the document]
Comment (vloison) [reply fadin/solver.py:337]: why ?

File: fadin/loss_and_gradient.py (hunk 1 above)
Comment (vloison) [reply fadin/loss_and_gradient.py]: Done :)

File: fadin/solver.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomMoral) [reply fadin/solver.py:336]: because the iteration of the optimizer is the next line.

here this function computes the gradient so renaming it would make it clearer IMO

File: fadin/tests/test_loss_and_gradient.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomMoral) [reply fadin/tests/test_loss_and_gradient.py:210]: great!

File: fadin/solver.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vloison) [reply fadin/solver.py]: makes sense, done :)

File: fadin/solver.py [hunk 5]
```diff
@@ -260,19 +258,42 @@ def fit(self, events, end_time):
...
+            )
+            self.baseline = baseline
+            self.alpha = alpha
+            # Set optimizer with moment_matching parameters
+            self.params_intens = [self.baseline, self.alpha]
+
+            for i in range(self.n_kernel_params):
+                kernel_param = kernel_params_init[i].float().clip(1e-4)
+                kernel_param.requires_grad_(True)
+                self.params_intens.append(kernel_param)
```
[the remark this answers is not in the document]
Comment (vloison) [reply fadin/solver.py:285]: Sounds good! Do the parameters in the random and given case need to be initialized in `__init__`, or can we call `init_params` once for all cases in `fit`?

File: fadin/utils/utils.py [hunk 6]
```diff
@@ -127,3 +128,114 @@ def check_random_state(seed):
         return seed
     raise ValueError('%r cannot be used to seed a numpy.random.RandomState'
                      ' instance' % seed)
+
+
+def momentmatching_kernel_nomark(solver, events, n_ground_events,
```
[the remark this answers is not in the document]
Comment (tomMoral) [reply fadin/utils/utils.py:133]: Call this `moment_matching.py`?

File: fadin/solver.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomMoral) [reply fadin/solver.py:285]: I would call once for all in `fit`

File: fadin/utils/utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vloison) [reply fadin/utils/utils.py]: I stuck with `init.py`, all possible initiaizations are done in this file.

File: fadin/init.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
+import torch
+import matplotlib.pyplot as plt
+
+
+def init_hawkes_params(solver, init_mode, events, n_ground_events, end_time):
+    """
+    Computes the initial Hawkes parameters for the FaDIn solver.
+    The function supports three modes of initialization:
```
Comment (tomMoral) [inline fadin/init.py:8]:
```suggestion
    Computes the initial Hawkes parameters for the FaDIn solver.

    The function supports three modes of initialization:
```

File: fadin/init.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    - 'random': Random initialization of parameters.
+    - 'moment_matching': Moment matching initialization of parameters.
+    - given: parameters are given by user.
+    ----------
```
Comment (tomMoral) [inline fadin/init.py:12]:
```suggestion

    Parameters
    ----------
```

File: fadin/init.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    events: list of array of size number of timestamps,
+        list size is self.n_dim.
+    n_ground_events : torch.tensor
+        Number of ground events for each dimension
+    end_time: float
+        End time of the events time series.
+
+    Returns:
+    --------
+    params_intens: list
```
Comment (tomMoral) [inline fadin/init.py:27]: Why not a dict?

File: fadin/init.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        kernel_param.requires_grad_(True)
+        params_intens.append(kernel_param)
+
+    return params_intens
+
+
+def momentmatching_kernel_nomark(solver, events, n_ground_events,
+                                 plot_delta=False, mode='max'):
+    """Moment matching initialization of kernel parameters. Implemented for
+    'truncated_gaussian' and 'raised_cosine' kernels.
```
Comment (tomMoral) [inline fadin/init.py:70]:
```suggestion
    """Moment matching initialization of kernel parameters.

    Implemented for 'truncated_gaussian' and 'raised_cosine' kernels.
```

File: fadin/init.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+                u = max(avg - std, 0)
+                kernel_params_init[0][i, j] = u
+            kernel_params_init[1][i, j] = std
+    return kernel_params_init
+
+
+def momentmatching_nomark(solver, events, n_ground_events, end_time,
+                          mode='max'):
+    """Moment matching initialization of baseline, alpha and kernel parameters.
+    $\\mu_i^s = \frac{\\#\\mathscr{F}^i_T}{(D+1)T} \forall i \\in [1, D]$
```
Comment (tomMoral) [inline fadin/init.py:156]:
```suggestion
    """Moment matching initialization of baseline, alpha and kernel parameters.

    $\\mu_i^s = \frac{\\#\\mathscr{F}^i_T}{(D+1)T} \forall i \\in [1, D]$
```

File: fadin/init.py [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+    $\\alpha_{i, j}^s = \\frac{1}{D+1} \\forall i,j \\in [1, D]$
+    Kernel parameters are initialized by `momentmatching_kernel`.
+    """
```
Comment (tomMoral) [inline fadin/init.py:159]: Add `Parameters` section.

File: fadin/init.py [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+    # Kernel parameters init
+    kernel_params_init = momentmatching_kernel_nomark(
+        solver, events, n_ground_events, mode=mode
+    )
+    return baseline, alpha, kernel_params_init
+
+
+def random_params(solver):
+    """Random initialization of baseline, alpha and kernel parameters.
+    Baseline and alpha are initialized with uniform distribution.
```
Comment (tomMoral) [inline fadin/init.py:178]:
```suggestion
    """Random initialization of baseline, alpha, and kernel parameters.

    Baseline and alpha are initialized with uniform distribution.
```

File: fadin/init.py [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+    Kernel parameters are initialized with uniform distribution.
+    """
```
Comment (tomMoral) [inline fadin/init.py:180]: Add `Parameters` section

File: fadin/solver.py [hunk 15]
```diff
@@ -148,72 +146,56 @@ def __init__(self, n_dim, kernel, init='random', optim_mask=None,
...
+        elif init == 'random':
+            self.init_mode = 'random'
+        else:
+            self.init_mode = init
+            assert 'baseline' in init.keys(), \
+                "baseline parameter must be given"
+            assert 'alpha' in init.keys(), \
+                "alpha parameter must be given"
+            assert 'kernel' in init.keys(), \
+                "kernel parameter must be given"
```
Comment (tomMoral) [inline fadin/solver.py:192]: This  should be mostly done in `inti_hawkes_params`

here, you should only have param validation:
```python
if isinstance(init, str):
    assert init in [....]
else:
     assert isinstance(init, dict) and have the right key.
```

File: fadin/init.py (hunk 9 above)
Comment (vloison) [reply fadin/init.py:30]: Great point, for params_intens` right?

File: fadin/init.py (hunk 14 above)
Comment (vloison) [reply fadin/init.py:226]: done :)

File: fadin/init.py (hunk 12 above)
Comment (vloison) [reply fadin/init.py:191]: done :)

File: fadin/init.py (hunk 9 above)
Comment (tomMoral) [reply fadin/init.py:31]: yes

Comment (tomMoral) [reply fadin/init.py:31]: ok, leave it like that for now then :)

File: fadin/solver.py (hunk 15 above)
Comment (vloison) [reply fadin/solver.py]: done in [commit 2f5b2a0](https://github.com/mind-inria/FaDIn/pull/16/commits/2f5b2a05ccf45dab43c6d0c639ddff7aa3865758) :)

File: fadin/solver.py [hunk 16]
```diff
@@ -413,7 +315,8 @@ def fit(self, events, end_time):
         return self
 
 
-def plot(solver, plotfig=False, bl_noise=False, title=None, ch_names=None, savefig=None):
+def plot(solver, plotfig=False, bl_noise=False, title=None, ch_names=None,
```
Comment (tomMoral) [inline fadin/solver.py:318]: This should be moved in a separate file in `utils` in a follow up PR.

File: fadin/init.py [hunk 17]
```diff
@@ -0,0 +1,259 @@
+import torch
+import matplotlib.pyplot as plt
```
Comment (tomMoral) [inline fadin/init.py:2]: Is this needed?

File: fadin/utils/utils.py [hunk 18]
```diff
@@ -34,6 +34,26 @@ def kernel_deriv_norm(function, grad_function_param, delta):
...
+        kernel_params_init.append(temp)
+        kernel_params_init.append(temp2)
+    elif kernel == 'truncated_exponential':
+        kernel_params_init.append(2 * torch.rand(n_dim, n_dim))
+    else:
+        raise NotImplementedError('kernel initial parameters of not \
+                                    implemented kernel have to be given')
+    return kernel_params_init
+
+
```
Comment (tomMoral) [inline fadin/utils/utils.py:56]: Is this still use somewhere?

**Review by tomMoral — COMMENTED** [review]
A few extra comments, when taken into account, feel free to merge.

File: fadin/init.py (hunk 17 above)
Comment (vloison) [reply fadin/init.py:2]: I think so because both imports are used in the file. Am I missing something?

File: fadin/utils/utils.py (hunk 18 above)
Comment (vloison) [reply fadin/utils/utils.py:56]: No, I'll delete it.

File: fadin/solver.py (hunk 16 above)
Comment (vloison) [reply fadin/solver.py:318]: Something like a `plot.py` file in the utils folder?

File: fadin/init.py [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+        mm_mode = init.split('_')[-1]
+        baseline, alpha, kernel_params_init = momentmatching_nomark(
+            solver,
+            events,
+            n_ground_events,
+            end_time,
+            mm_mode
+        )
+    elif init == 'random':
+        baseline, alpha, kernel_params_init = random_params(solver)
```
[the remark this answers is not in the document]
Comment (vloison) [reply fadin/init.py:55]: I added it to the to-do list in issue https://github.com/mind-inria/FaDIn/issues/17 .

File: fadin/solver.py (hunk 16 above)
Comment (vloison) [reply fadin/solver.py]: Done in commit [a7c924e](https://github.com/mind-inria/FaDIn/pull/16/commits/a7c924e3ffd3909338f6b534c4164438636701d8)

File: fadin/init.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (tomMoral) [reply fadin/init.py:55]: nice!
